docs/design/2024-05-10-extension-authentication-plugin.md

```diff
@@ -0,0 +1,231 @@
+    // ctx: context passed in from caller
+    // authConn: interface for the plugin to communicate with the client
+    PasswordReader() func(ctx context.Context, authConn conn.AuthConn) ([]byte, error)
+    
+    // Authenticate is called when a client connects to the server as a user and the server authenticates the user.
+    // If an error is returned, the login attempt fails, otherwise it succeeds.
+    // authUser: the username in the connect attempt
+    // storedPwd: the user's password stored in mysql.user table
+    // inputPwd (authentication): the user's password passed in from the connection attempt in bytes (return value of `plugin.PasswordReader()(ctx, authConn, resp)` if `plugin.PasswordReader() != nil`)
+    // salt: randomly generated salt for the current connection
+    // connState: the TLS connection state (contains the TLS certificate) if client is using TLS. It will be nil if the client is not using TLS
+    // authConn: interface for the plugin to communicate with the client
+    Authenticate(authUser string, storedPwd string, inputPwd []byte, salt []byte, connState *tls.ConnectionState, authConn conn.AuthConn) error
+    
+    // EncodePassword is a function for user to implement customized ways to encode the password (e.g. hash/salt/clear-text). The returned string will be stored as the encoded password in the mysql.user table.
+    // If the input password is considered as invalid, this should return an error.
+    // pwd: User's input password in CREATE/ALTER USER statements in clear-text
+    EncodePassword(pwd string) (string, bool)
+    
+    // IsValidPasswordHash checks if the password hash stored in the mysql.user table is valid.
+    // This is called when retrieving an existing user to make sure the password stored is valid and not modified.
+    // pwdHash: hash of the password stored in the internal user table
+    IsValidPasswordHash(pwdHash string) bool
+    
+    // VerifyPrivilege is called for each user queries, and serves as an extra check for privileges for the user.
+    // It will only be executed if the user has already been granted the privilege in SQL layer.
+    // Returns true if user has the requested privilege.
+    // activeRoles: list of active MySQL roles for the current user
+    // user: current user's name
+    // host: the host that the user is connecting from
+    // db: the database to check for privilege
+    // table: the table to check for privilege
+    // column: the column to check for privilege (currently just a placeholder in TiDB as column-level privilege is not supported by TiDB yet)
+    // priv: the privilege type of the SQL statement that will be executed
+    // connState: the TLS connection state (contains the TLS certificate) if client is using TLS. It will be nil if the client is not using TLS
+    VerifyPrivilege(activeRoles []*auth.RoleIdentity, user, host, db, table, column string, priv mysql.PrivilegeType, connState *tls.ConnectionState) bool
```

Comment:
If two connections are identified with the same user and the same plugin, is this method expected to return the same value when the input arguments are also the same?

Reply:
That would depend on the implementation of it. The developer could choose to return different values based on the use case.